src/statsmodels/formula.jl

```diff
@@ -234,34 +240,149 @@ function dropUnusedLevels!(da::PooledDataArray)
+    non_redundants = check_non_redundancy(trms, df)
+
+    ModelFrame(df, trms, msng, evaledContrasts, non_redundants)
 end
 
-ModelFrame(f::Formula, d::AbstractDataFrame) = ModelFrame(Terms(f), d)
-ModelFrame(ex::Expr, d::AbstractDataFrame) = ModelFrame(Formula(ex), d)
+ModelFrame(f::Formula, d::AbstractDataFrame; kwargs...) = ModelFrame(Terms(f), d; kwargs...)
+ModelFrame(ex::Expr, d::AbstractDataFrame; kwargs...) = ModelFrame(Formula(ex), d; kwargs...)
+
+## modify contrasts in place
+function contrast!(mf::ModelFrame, new_contrasts::Dict)
+    new_contrasts = [ col => evaluateContrasts(contr, mf.df[col])
+                      for (col, contr) in filter((k,v)->haskey(mf.df, k), new_contrasts) ]
+                      
+    mf.contrasts = merge(mf.contrasts, new_contrasts)
+    return mf
+end
+contrast!(mf::ModelFrame; kwargs...) = contrast!(mf, Dict(kwargs))
 
 function StatsBase.model_response(mf::ModelFrame)
     mf.terms.response || error("Model formula one-sided")
     convert(Array, mf.df[round(Bool, mf.terms.factors[:, 1])][:, 1])
 end
 
-function contr_treatment(n::Integer, contrasts::Bool, sparse::Bool, base::Integer)
-    if n < 2 error("not enought degrees of freedom to define contrasts") end
-    contr = sparse ? speye(n) : eye(n) .== 1.
-    if !contrasts return contr end
-    if !(1 <= base <= n) error("base = $base is not allowed for n = $n") end
-    contr[:,vcat(1:(base-1),(base+1):end)]
+modelmat_cols(v::DataVector) = convert(Vector{Float64}, v.data)
+modelmat_cols(v::Vector) = convert(Vector{Float64}, v)
+## construct model matrix columns from model frame + name (checks for contrasts)
+function modelmat_cols(name::Symbol, mf::ModelFrame; non_redundant::Bool = false)
+    if haskey(mf.contrasts, name)
+        modelmat_cols(mf.df[name],
+             non_redundant ? promote_contrast(mf.contrasts[name]) : mf.contrasts[name])
+    else
+        modelmat_cols(mf.df[name])
+    end
 end
-contr_treatment(n::Integer,contrasts::Bool,sparse::Bool) = contr_treatment(n,contrasts,sparse,1)
-contr_treatment(n::Integer,contrasts::Bool) = contr_treatment(n,contrasts,false,1)
-contr_treatment(n::Integer) = contr_treatment(n,true,false,1)
-cols(v::PooledDataVector) = contr_treatment(length(v.pool))[v.refs,:]
-cols(v::DataVector) = convert(Vector{Float64}, v.data)
-cols(v::Vector) = convert(Vector{Float64}, v)
+
+"""
+    modelmat_cols(v::PooledDataVector, contrast::ContrastsMatrix)
+
+Construct `ModelMatrix` columns based on specified contrasts, ensuring that
+levels align properly.
+"""
+function modelmat_cols(v::PooledDataVector, contrast::ContrastsMatrix)
+    ## make sure the levels of the contrast matrix and the categorical data
+    ## are the same by constructing a re-indexing vector. Indexing into
+    ## reindex with v.refs will give the corresponding row number of the
+    ## contrast matrix
+    reindex = [findfirst(contrast.levels, l) for l in levels(v)]
+    return contrast.matrix[reindex[v.refs], :]
```

Comment:
So this may change the order of appearance of columns and coefficients? Doesn't sound like a good idea, especially for Helmert contrasts.